TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/TrafficReplayer.java

```diff
@@ -164,30 +169,43 @@ private void runReplayWithIOStreams(Duration observedPacketConnectionTimeout,
                             }
-                            var resultantCf = requestFutureMap.get(rrPair.requestData)
-                                    .handle((summary, t) -> {
-                                        try {
-                                            var rval = packageAndWriteResponse(tupleWriter, rrPair, summary, t);
-                                            successCount.incrementAndGet();
-                                            return rval;
-                                        } catch (Exception e) {
-                                            exceptionCount.incrementAndGet();
-                                            throw e;
-                                        }
-                                    })
-                                    .whenComplete((v,t) -> requestToFinalWorkFuturesMap.remove(rrPair.requestData));
-                            requestToFinalWorkFuturesMap.put(rrPair.requestData, resultantCf);
+                            DiagnosticTrackableCompletableFuture<String,AggregatedTransformedResponse> resultantCf =
+                                    requestFutureMap.get(rrPair.requestData)
+                                            .map(f->
+                                                    f.handle((summary, t) -> {
+                                                        try {
+                                                            AggregatedTransformedResponse rval =
+                                                                    packageAndWriteResponse(tupleWriter, rrPair, summary, t);
+                                                            successCount.incrementAndGet();
+                                                            return rval;
+                                                        } catch (Exception e) {
+                                                            exceptionCount.incrementAndGet();
+                                                            throw e;
+                                                        } finally {
+                                                            requestToFinalWorkFuturesMap.remove(rrPair.requestData);
+                                                            log.error("removed rrPair.requestData to requestToFinalWorkFuturesMap for " +
+                                                                    rrPair.connectionId);
+                                                        }
+                                                    }), ()->"TrafficReplayer.runReplayWithIOStreams.progressTracker");
+                            if (!resultantCf.future.isDone()) {
+                                log.error("Adding " + rrPair.connectionId + " to requestToFinalWorkFuturesMap");
+                                requestToFinalWorkFuturesMap.put(rrPair.requestData, resultantCf);
+                                if (resultantCf.future.isDone()) {
```

Comment:
What is point of this if statement, to potentially catch futures that have completed since lines 203 and 204?

Reply:
Yep - to deal with race conditions.  I don't think that I have a unit test to deal with this, but I did notice that when dealing w/ my static input file, this happened to a number of entries that just lingered in and were wasting space in the allRemainingWork checks at the end (they were complete, so it wasn't a big deal, but it created more noise for when I was debugging)